File: leetcode/20. Valid Parentheses/solution.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> st;
        for (int i = 0; i < s.size(); i++)
        {
            if (st.empty())
            {
                st.push(s[i]);
            }
            else
            {
                char t = st.top();
                if ((t == '{' && s[i] == '}') || (t == '[' && s[i] == ']') || (t == '(' && s[i] == ')'))
                {
                    st.pop();
                }
                else
                    st.push(s[i]);
            }
            // if (!st.empty())
            //     cout << st.top() << endl;
            // st.empty() ? cout << "empty\n" : cout << "not empty\n";
        }
        bool ans;
        st.empty() ? ans = true : ans = false;
        return ans;
    }
};
int main()
{
    string s;
    cin >> s;
    Solution S;
    cout << S.isValid(s) << endl;
    return 0;
}

File: leetcode/20. Valid Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for i in range(len(s)):
            if stack == []:
                stack.append(s[i])
            else:
                top = stack[-1]
                if (top == '(' and s[i] == ')') or (top == '{' and s[i] == '}') or (top == '[' and s[i] == ']'):
                    stack.pop()
                else:
                    stack.append(s[i])
        return stack == []

sol = Solution()
s = input()
print(sol.isValid(s))